### PR TITLE
[Op] [TVM] update reduce_scatter

### PR DESCRIPTION
**Pull request by StrongSpoon**
## Description ##
add argument rank_list to support reduce_scatter among sectional ranks
bash ci/task_lint.sh passed

## Checklist ##

- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

cc @awslabs/raf-reviewer

### CODE REVIEW

**Comment by hgt312** [comment]
@comaniac @yzhliu @zachzzc @hzfan

File: tests/python/distributed/test_collective_communication.py [hunk 1]
```diff
@@ -305,6 +306,15 @@ def forward(self, x, y):
...
     m_out = run_model(model, [m_x, m_y], device)
+    for group in rank_list:
+        if rank in group:
+            ones = np.ones(shape=(4, 4), dtype="float32")
+            n_out = ones * sum(np.array(group) + 1)
+            if rank % 2 == 0:
+                check(m_out, target_x)
+            else:
+                check(m_out, -target_x)
+    """
```
Comment (hgt312) [inline tests/python/distributed/test_collective_communication.py:317]: why comment these?

**Review by hgt312 — CHANGES_REQUESTED** [review]

File: tests/python/distributed/test_collective_communication.py [hunk 2]
```diff
@@ -280,17 +280,18 @@ def forward(self, x1, x2):
             check(y, target_y)
 
 
-@pytest.mark.skipif(skip_dist_test(min_rank_num=2, require_exact_rank=True), reason=SKIP_REASON)
+@pytest.mark.skipif(skip_dist_test(min_rank_num=4, require_exact_rank=True), reason=SKIP_REASON)
 @pytest.mark.parametrize("computation", ["sum", "prod", "min", "max"])
-def test_reduce_scatter(computation):
+@pytest.mark.parametrize("rank_list", [[[0, 1], [2, 3]]]) # invalid rank_list [[1, 2, 3]]
```
[the remark this answers is not in the document]
Comment (hgt312) [reply tests/python/distributed/test_collective_communication.py:285]: Looks that op's type infer is not changed, so `rank_list=[[1, 2, 3]]` failed.

**Comment by hgt312** [comment]
Now RAF's `reduce_scatter` follows PyTorch's semantic, which requires size of input tensor list being equal to communicator's size then outout size is equal to each input.

It has some conflict with `rank_list`.

For example, if we are on a machine with 4 gpus, with `reduce_scatter(x = [inp0, inp1, inp2, inp3], rank_list = [[0, 1], [2, 3]])`. Under current impl, we will get output's size double to each input. According to pytorch's dist op design, we only want inp0&inp1 participate computation of rank0&rank1 and inp2&inp3 participate computation of rank2&rank3.

It depends on how we concat the inputs (if follow the rank_list). If we choose to make concat's behavior follow the rank list, I think we can change `reduce_scatter`'s impl to make it a pure NCCL reduce-scatter, and make concat in the python wrapper, which is easy to implement.

@comaniac @yzhliu @zachzzc @hzfan @zhen-jia

**Comment by comaniac** [comment]
@hgt312 is it similar to the issue that @Tonny-Gu faced previously? I mean the semantic differences of collective ops between PyTorch and GSPMD. If so, it would be better to have a design document to discuss all these semantic gaps and make decisions at once.

**Comment by Tonny-Gu** [comment]
There should not be a huge semantic gap between RAF and other frameworks.

In PyTorch, the design of AllReduce API is based on what they call `ProcessGroup`, which is a bit similar to our `Communicator`. And the API looks like this,

[Ref: https://pytorch.org/docs/stable/distributed.html](https://pytorch.org/docs/stable/distributed.html)

```python
torch.distributed.all_gather(tensor_list, tensor, group=None, async_op=False)
```

where `group` is the process group to work on. If None, the default process group will be used.

For example, if a PyTorch process group refers to Rank 0-3, the equivalent expression of RAF should be,

```python
raf.allgater(rank_list=[[0, 1, 2, 3]])
```

Conversely, since we allow to write `rank_list` like this,

```python
raf.allgater(rank_list=[[0, 1], [2, 3]])
```

This expression should equal to,

```python
torch.all_gather(group=G1)
torch.all_gather(group=G2)
```

where `G1` and `G2` refer to Rank 0-1 and 2-3 respectively.

In JAX, there is a similar concept named `axis_index_groups`, and its semantics exactly matches our `rank_list`. Here is an example provided by the official,

[Ref: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.all_gather.html#jax.lax.all_gather](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.all_gather.html#jax.lax.all_gather)

```python
x = np.arange(16).reshape(4, 4)
print(x)
  [[ 0  1  2  3] # On Device 0
   [ 4  5  6  7] # On Device 1
   [ 8  9 10 11] # On Device 2
   [12 13 14 15]] # On Device 3
def f(x):
  return jax.lax.all_gather(
      x, 'i', axis_index_groups=[[0, 2], [3, 1]])
y = jax.pmap(f, axis_name='i')(x)
print(y)
[[[ 0  1  2  3]
  [ 8  9 10 11]] # On Device 0
 [[12 13 14 15]
  [ 4  5  6  7]] # On Device 1
 [[ 0  1  2  3] 
  [ 8  9 10 11]] # On Device 2
 [[12 13 14 15]
  [ 4  5  6  7]]] # On Device 3
```

And if we execute `y = allreduce(x, rank_list=[[0, 2], [3, 1]])`, we could expect the following result appeared on each rank.

| Var | Rank 0 | Rank 1 | Rank 2 | Rank 3 |
| --- | --- | --- | --- | --- |
| x | [0, 1, 2, 3] | [4, 5, 6, 7] | [8, 9, 10, 11] | [12, 13, 14, 15] |
| y | [[0..3], [8..11]] | [[12..15], [4..7]] | [[0..3], [8..11]] | [[12..15], [4..7]] |

In PyTorch, this can be achieved with two function calls and two process groups referring to Rank 0, 2 and Rank 3, 1.

The reasons why we design `rank_list` in this way are,

- Specifying `rank_list` may implicitly create a new communicator. And the creation of a communicator is a collective operation, which requires all ranks to participate. Therefore, we hope to let as many ranks as possible get involved and perform real computation at a time. 
- Simplify the expression. We don't have to invoke collective operators multiple times to perform the same operation on each process group one by one.

```python
# Originally
y1 = allreduce(x, rank_list=[0, 1])
y2 = allreduce(x, rank_list=[2, 3])
if rank in (0, 1):
  y = y1
elif rank in (2, 3):
  y = y2

# Now
y = allreduce(x, rank_list=[[0, 1], [2, 3]])
```

For now, we assume our IR is written in the SPMD style, and we allow the rank not to appear in `rank_list` (unlike JAX). And the rank not in `rank_list` will work in the standalone mode instead of skipping this Op Call. By doing so, we can guarantee those unrelated ranks will get some reasonable results (like placeholders). Note that we still need to reference these results somewhere to prevent the Op Call from being removed by the compiler.

Additionally, the sharding specification should be a different system in contrast to `rank_list`. When we lower an Op Call with `ShardOpCallAttrs` to an Op Call without this attribute, `ShardSpec` will be converted into `rank_list` as well as other arguments like `axis`.

**Comment by comaniac** [comment]
@Tonny-Gu thanks for the explanation. These semantic gaps should be documented in the sharding design doc to support the RAF sharding system design.

**Comment by hgt312** [comment]
What @Tonny-Gu post is not related to the issue. But actually we need to discuss about the "standalone" mode vs skip.

The issue I want to point out is not rank_list or GSPMD, but dist ops' **input tensor list with rank list**, which is about **how input(s)/output(s) get packed/unpacked**, **how the semantics of an operator is defined**.

Now the design of RAF's dist ops is not uniform. I think we can do some survey about the exist DL compilers/frameworks, and have a uniform design for all dist ops.

------

Especially for reduce_scatter:
```python
if current_rank in [0, 1]:
    torch.distributed.reduce_scatter(out_buffer0, input_list=[inp0, inp1], group=G1)  # G1 - rank 0&1
if current_rank in [2, 3]:
    torch.distributed.reduce_scatter(out_buffer1, input_list=[inp2, inp3], group=G2)  # G2 - rank 2&3
```
is equal to
```python
if current_rank in [0, 1]:
    out0 = raf.reduce_scatter([inp0, inp1], rank_list=[[0, 1]])
if current_rank in [2, 3]:
    out1 = raf.reduce_scatter([inp2, inp3], rank_list=[[2, 3]])
```
but is not equal to something like
```python
# out will be quite different with out0/out1
# Do we want `[out0, out1] = raf.reduce_scatter([inp0, inp1, inp2, inp3], rank_list=[[0, 1], [2, 3]])`?
out = raf.reduce_scatter([inp0, inp1, inp2, inp3], rank_list=[[0, 1], [2, 3]])
```

**Comment by Tonny-Gu** [comment]
@hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` has no relationship with `inp1`?

Fig1

**Comment by hgt312** [comment]
> @hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` have no relationship with `inp1`?
> 
> Fig1

inp0 refer to the first input tensor on each rank

**Comment by comaniac** [comment]
Per offline discussion, we will have a design doc driven by @StrongSpoon to enhance collective ops.

**Comment by Tonny-Gu** [comment]
> > @hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` have no relationship with `inp1`?
> > Fig1
> 
> inp0 refer to the first input tensor on each rank

And `inp0` is evenly partitioned into several data shards, like `in0`, `in1`, `in2`, `in3` in the figure above. Is that true?

**Comment by Tonny-Gu** [comment]
> > @hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` have no relationship with `inp1`?
> > Fig1
> 
> inp0 refer to the first input tensor on each rank

So you mean `inp0` refer to `in0`, `in1`, `in2`, `in3`. Correct?

**Comment by hgt312** [comment]
> > > @hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` have no relationship with `inp1`?
> > > Fig1
> > 
> > 
> > inp0 refer to the first input tensor on each rank
> 
> So you mean `inp0` refer to `in0`, `in1`, `in2`, `in3`. Correct?

For this picture, `concat([inp0, inp1, inp2, inp3])` refer to `in0`, `in1`, `in2`, `in3`.

**Comment by Tonny-Gu** [comment]
> > > > @hgt312 Does your `inp0` refer to `in0` on the rank 0 in the figure below? Or does `inp0` have no relationship with `inp1`?
> > > > Fig1
> > > 
> > > 
> > > inp0 refer to the first input tensor on each rank
> > 
> > 
> > So you mean `inp0` refer to `in0`, `in1`, `in2`, `in3`. Correct?
> 
> For this picture, `concat([inp0, inp1, inp2, inp3])` refer to `in0`, `in1`, `in2`, `in3`.

I see the dilemma. I personally prefer the design that `raf.reduce_scatter([inp0, inp1, inp2, inp3], rank_list=[[0, 1], [2, 3]])` will generate four output tensors, and this usage should be equivalent to

```python
# (out0, out1, out2, out3) = raf.reduce_scatter([inp0, inp1, inp2, inp3], rank_list=[[0, 1], [2, 3]])

if rank in [0, 1]:
  out0 = raf.reduce_scatter([inp0], rank_list=[[0, 1]])
else if rank in [2, 3]:
  out0 = raf.reduce_scatter([inp0], rank_list=[[2, 3]])
....
if rank in [0, 1]:
  out3 = raf.reduce_scatter([inp3], rank_list=[[0, 1]])
else if rank in [2, 3]:
  out3 = raf.reduce_scatter([inp3], rank_list=[[2, 3]])
```

That should align to our current design.

**Comment by Tonny-Gu** [comment]
> ```python
> # it's not runnable due to "standalone" mentioned above
> if current_rank in [0, 1]:
>     out0 = raf.reduce_scatter([inp0, inp1], rank_list=[[0, 1]])
> if current_rank in [2, 3]:
>     out1 = raf.reduce_scatter([inp2, inp3], rank_list=[[2, 3]])
> ```

This code should be able to run, but it is not efficient due to the current design. We indeed could let the rank not in `rank_list` skip the irrelevant code.

**Comment by zhen-jia** [comment]
Looking forward to the design doc.  The current dist ops were implemented by different people at different time. Some of the interface may out of data. We need to refactor and unify them.

**Review by hgt312 — APPROVED** [review]
LGTM

**Comment by hgt312** [comment]
Thanks @StrongSpoon @comaniac @Tonny-Gu
